### PR TITLE
WIP Initial work on integrating the Zone library as an entity+plugin combo

### PR DESCRIPTION
**Pull request by Kazanir**
Needs the actual forms for each zone member type created and integrated in a way that loads the containing entity and saves it properly in the form submit. The rest is mostly there.

### CODE REVIEW

File: address.links.menu.yml [hunk 1]
```diff
@@ -3,3 +3,9 @@ entity.address_format.collection:
   parent: system.admin_config_regional
   description: 'Administer the address formats for each country.'
   route_name: entity.address_format.collection
+
+entity.zone.collection:
+  title: 'Zones'
+  route_name: entity.zone.collection
+  parent: system.admin_config_regional
+  description: 'Administer the zone entity defined by the addressmodule.'
```
Comment (bojanz) [inline address.links.menu.yml]: Let's do a nicer description. Something as simple as "Administer zones"? or "Administer territorial groupings"?

File: address.routing.yml [hunk 2]
```diff
@@ -83,3 +83,64 @@ entity.subdivision.delete_form:
...
+  path: '/admin/config/regional/zones/{zone}/delete'
+  defaults:
+    _entity_form: 'zone.delete'
+    _title: 'Delete a zone'
+  options:
+    _admin_route: TRUE
+  requirements:
+    _permission: 'administer zones'
+
+entity.zonemember.edit_form:
```
Comment (bojanz) [inline address.routing.yml]: Zone members aren't entities, so these routes have the wrong pattern.
"address.zone_member.edit_form"?

**Comment by bojanz** [comment]
So, "zonemember" should be "zone_member" everywhere.

**Comment by Kazanir** [comment]
Ugh underscores ><

File: address.services.yml [hunk 3]
```diff
@@ -11,6 +11,10 @@ services:
     class: Drupal\address\CountryRepository
     arguments: ['@cache.data', '@language_manager']
 
+  plugin.manager.address.zone_member:
```
Comment (bojanz) [inline address.services.yml]: Let's move this service to the end (since it starts with a p)

File: src/Entity/Zone.php [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+   * @var int
+   */
+  protected $priority;
+
+  /**
+   * Zone members.
+   *
+   * @var array
+   */
+  protected $members = array();
```
Comment (bojanz) [inline src/Entity/Zone.php]: array(); should be [];

File: src/Entity/Zone.php [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+   * @var \Drupal\Core\Plugin\DefaultLazyPluginCollection
+   */
+  protected $membersCollection;
+
+  /**
+   * Returns the string representation of the zone.
+   *
+   * @return string
+   */
+  public function __toString()
```
Comment (bojanz) [inline src/Entity/Zone.php]: All of these methods have the opening brace on the wrong line.

File: src/Entity/Zone.php [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getMembers()
+  {
+
+    if (!$this->membersCollection) {
+      if (!is_array($this->members)) { $this->members = []; }
```
Comment (bojanz) [inline src/Entity/Zone.php]: Everything is ugly in this method :)

File: src/Entity/Zone.php [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function match(AddressInterface $address)
+  {
+    foreach ($this->membersCollection as $member) {
+      if ($member->match($address)) {
+        return true;
```
Comment (bojanz) [inline src/Entity/Zone.php]: true and false need be TRUE and FALSE;

File: src/Form/ZoneDeleteForm.php [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+namespace Drupal\address\Zone\Form;
+
+use Drupal\Core\Entity\EntityConfirmFormBase;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Url;
+
+/**
+ * Builds the form to delete a zone.
+ */
+class ZoneDeleteForm extends EntityConfirmFormBase {
```
Comment (bojanz) [inline src/Form/ZoneDeleteForm.php]: Extend EntityDeleteForm instead, remove redundant code.

File: src/Form/ZoneForm.php [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+   * The zone member plugin manager.
+   *
+   * @var \Drupal\address\ZoneMemberManager
+   */
+  protected $zoneMemberManager;
+
+  /**
+   * Creates a ZoneForm instance.
+   *
+   * @param \Drupal\Core\Entity\EntityStorageInterface $zone_storage
```
Comment (bojanz) [inline src/Form/ZoneForm.php]: Let's inject the $entityManager into the constructor, then do $this->zoneStorage = $entityManager->getStorage('zone'); It matches how we're dojng it in Commerce now.

File: src/Form/ZoneForm.php [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+   *   The zone storage.
+   */
+  public function __construct(EntityStorageInterface $zone_storage, ZoneMemberManager $zone_member_manager) {
+    $this->zoneStorage = $zone_storage;
```
Comment (bojanz) [inline src/Form/ZoneForm.php]: All variables must be camelCase.

File: src/Form/ZoneForm.php [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function form(array $form, FormStateInterface $form_state) {
+    $user_input = $form_state->getUserInput();
+    $zone = $this->entity;
+
+    $form['name'] = array(
```
Comment (bojanz) [inline src/Form/ZoneForm.php]: Old array syntax. Array( should be [

File: src/Form/ZoneForm.php [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+      '#attributes' => array(
+        'id' => 'zone-zone-members',
+      ),
+      '#empty' => t('There are currently no zone members in this zone. Add one by selecting an option below.'),
+      // Render zone members below parent elements.
+      '#weight' => 5,
+    );
+
+    foreach ($this->entity->getMembers() as $key => $member) {
+      $member_form =& $form['members'][$key];
```
Comment (bojanz) [inline src/Form/ZoneForm.php]: Looks like the member form could be built in a different method.

File: src/Form/ZoneForm.php [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+    $id = $element['#value'];
+    if (!preg_match('/[a-z_]+/', $id)) {
+      $form_state->setError($element, $this->t('The machine name must be in lowercase, underscore-separated letters only.'));
+    }
+    elseif ($zone->isNew()) {
+      $loaded_zones = $this->zoneStorage->loadByProperties(array(
+        'id' => $id,
+      ));
+      if ($loaded_zones) {
+        $form_state->setError($element, $this->t('The machine name is already in use.'));
```
Comment (bojanz) [inline src/Form/ZoneForm.php]: Why aren't we using #exists on the form element for this?

File: src/Form/ZoneMemberDeleteForm.php [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    $this->zone = $zone;
+    $this->zoneMember = $this->zone->getMember($member);
+
+    return parent::buildForm($form, $form_state);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormSTateInterface $form_state) {
```
Comment (bojanz) [inline src/Form/ZoneMemberDeleteForm.php]: Typo: FormSTateInterface

File: src/Form/ZoneMemberForm.php [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+
+
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    // The image effect configuration is stored in the 'data' key in the form,
+    // pass that through for validation.
+    $effect_data = (new FormState())->setValues($form_state->getValue('data'));
```
Comment (bojanz) [inline src/Form/ZoneMemberForm.php]: Image effects?

File: src/ZoneMemberInterface.php [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+/**
+ * @file
+ * Contains \Drupal\address\ZoneMemberInterface.
+ */
+
+namespace Drupal\address;
+
+use Drupal\Component\Plugin\ConfigurablePluginInterface;
+use Drupal\Component\Plugin\PluginInspectionInterface;
+use CommerceGuys\Zone\Model\ZoneMemberInterface as LibZoneMemberInterface;
```
Comment (bojanz) [inline src/ZoneMemberInterface.php]: Let's call it ExternalZoneMemberInterface.

File: address.routing.yml [hunk 17]
```diff
@@ -29,3 +29,64 @@ entity.address_format.delete_form:
...
   requirements:
     _entity_access: 'address_format.delete'
+
+entity.zone.collection:
+  path: '/admin/config/regional/zones'
+  defaults:
+    _entity_list: 'zone'
+    _title: 'Zones'
+  options:
+    _admin_route: TRUE
```
Comment (bojanz) [inline address.routing.yml]: You don't need _admin_route if the path starts with "/admin", it's automatic in that case.

File: address.permissions.yml [hunk 18]
```diff
@@ -3,3 +3,8 @@
   description: 'Allows users to create, update, and delete address format entities.'
   'restrict access': TRUE
 
+'administer zones':
+  title: 'Administer addressing zones'
+  description: 'Allows users to create zones, or definitions for a territorial grouping of addresses.'
+  'restrict access': TRUE
```
Comment (bojanz) [inline address.permissions.yml]: restrict access? Which security implications does this UI have?

File: config/schema/address.schema.yml [hunk 19]
```diff
@@ -59,3 +59,63 @@ field.widget.settings.address_default:
...
+    locality:
+      type: string
+      label: 'Locality'
+    dependentLocality:
+      type: string
+      label: 'Code'
+    name:
+      type: label
+      label: 'Name'
+      translatable: true
```
Comment (bojanz) [inline config/schema/address.schema.yml]: type: label implies translatable: true, it doesn't need to be explicitly specified.

File: src/Form/ZoneDeleteForm.php [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+class ZoneDeleteForm extends EntityDeleteForm {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    try {
+      $this->entity->delete();
+      $form_state->setRedirectUrl($this->getCancelUrl());
+      drupal_set_message($this->t('Zone %label has been deleted.', array('%label' => $this->entity->label())));
```
Comment (bojanz) [inline src/Form/ZoneDeleteForm.php]: New code should not be using array(), let's use [] (short array syntax) instead.

File: src/Form/ZoneForm.php [hunk 21]
```diff
@@ -0,0 +1,288 @@
...
+      '#title' => $this->t('Weight for new member'),
+      '#title_display' => 'invisible',
+      '#default_value' => count($this->entity->getMembers()) + 1,
+      '#attributes' => ['class' => ['zonemember-order-weight']],
+    ];
+    $form['members']['new']['operations'] = [
+      'data' => [],
+    ];
+
+
```
Comment (bojanz) [inline src/Form/ZoneForm.php]: Why the so many empty rows?

File: src/Form/ZoneForm.php [hunk 22]
```diff
@@ -0,0 +1,288 @@
...
+  }
+
+  /**
+   * Validates the id field.
+   */
+  public function validateId(array $element, FormStateInterface $form_state, array $form) {
+    $zone = $this->getEntity();
+    $id = $element['#value'];
+    if (!preg_match('/[a-z_]+/', $id)) {
+      $form_state->setError($element, $this->t('The machine name must be in lowercase, underscore-separated letters only.'));
```
Comment (bojanz) [inline src/Form/ZoneForm.php]: This part can now be removed, you've defined the pattern on the machine name element?

File: src/Form/ZoneForm.php [hunk 23]
```diff
@@ -0,0 +1,288 @@
...
+    }
+    elseif ($zone->isNew()) {
+      $loaded_zones = $this->zoneStorage->loadByProperties([
+        'id' => $id,
+      ]);
+      if ($loaded_zones) {
+        $form_state->setError($element, $this->t('The machine name is already in use.'));
```
Comment (bojanz) [inline src/Form/ZoneForm.php]: You need to use the "exists" key on the machine name, killing the validateId method completely.

File: src/Form/ZoneForm.php (hunk 21 above)
Comment (Kazanir) [inline src/Form/ZoneForm.php]: Some combination of readability and it not being done yet :P

File: src/Form/ZoneForm.php [hunk 24]
```diff
@@ -0,0 +1,288 @@
...
+      drupal_set_message($this->t('The %label zone was not saved.', [
+        '%label' => $zone->label()
+      ]), 'error');
+      $this->logger('address.zone')->error($e);
+      $form_state->setRebuild();
+    }
+  }
+
+  /**
+   *
```
Comment (bojanz) [inline src/Form/ZoneForm.php]: Needs phpdoc.

File: address.permissions.yml (hunk 18 above)
Comment (Kazanir) [inline address.permissions.yml]: Well presumably for the same reason we (used to) warn about other Commerce permissions in that vein -- if someone can control zones then they can mess with how the store operates, like most "administer xyz" permissions.

File: src/Form/ZoneDeleteForm.php (hunk 20 above)
Comment (Kazanir) [inline src/Form/ZoneDeleteForm.php]: Yeah I am fixing those as I find them; I'll grep before all is said and done

File: src/Plugin/ZoneMember/ZoneMemberCountry.php [hunk 25]
```diff
@@ -0,0 +1,27 @@
...
+ * Matches a country, its subdivisions, and its postal codes.
+ *
+ * @ZoneMember(
+ *   id = "zone_member_country",
+ *   label = @Translation("Zone Member: Country"),
+ *   description = @Translation("Matches a country, its subdivisions, and its
+ *   postal codes.")
+ * )
+ */
+class ZoneMemberCountry extends Lib\ZoneMember implements ZoneMemberInterface {
```
Comment (bojanz) [inline src/Plugin/ZoneMember/ZoneMemberCountry.php]: Lib\ZoneMember is weird, let's just add a use statement for the actual class.

File: src/Form/ZoneDeleteForm.php (hunk 20 above)
Comment (bojanz) [inline src/Form/ZoneDeleteForm.php]: These two are the last ones :)

**Comment by bojanz** [comment]
Most of this has been committed now. Only the country plugin remains in progress.

**Pull request closed without merging**